### PR TITLE
Add SBT Registry

### LINKED ISSUE #1081: Create New zkSBT MintTypes Without a Runtime Upgrade
Right now to add new mint type we need update this enum: https://github.com/Manta-Network/Manta/blob/ceb9e46cd53b77eb914ba6c17452fc238bc3a28f/pallets/manta-sbt/src/lib.rs#L134

We should add an `MintType` registry to add new zkSBT `MintTypes` onchain dynamically. Migrating existing storage will be necessary, I think doing it lazily using `on_idle` hook is best way to go due to number of existing zkSBT mints already completed

### PR DESCRIPTION
**Pull request by ferrell-code**
## Description

Adds a `MintRegistry` and identifies mint types using `u32`. 

Migrates storage to new scheme using on_idle hook.

Manually bumps `NextMintId` to 3 using `on_runtime_upgrade`

closes: #1081 

---

Before we can approve this PR for merge, please make sure that **all** the following items have been checked off:
- [x] Connected to an issue with discussion and accepted design using zenhub "Connect issue" button below
- [x] Added **one** label out of the `L-` group to this PR
- [x] Added **one or more** labels from the `A-` and `C-` groups to this PR
- [x] Explicitly labelled `A-calamari`, `A-dolphin` and/or `A-manta` if your changes are meant for/impact either of these (CI depends on it)
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

Situational Notes:
- If adding functionality, write unit tests!
- If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: runtime/calamari/src/migrations/mod.rs [hunk 1]
```diff
@@ -15,5 +15,6 @@
 // along with Manta.  If not, see <http://www.gnu.org/licenses/>.
 
 pub mod asset_id;
+pub mod init_sbt_counter;
```
Comment (ghzlatarev) [inline runtime/calamari/src/migrations/mod.rs:18]: what do you guys think about commenting out the executed migrations, so we dont have to maintain

File: pallets/manta-sbt/src/tests.rs [hunk 2]
```diff
@@ -450,25 +458,28 @@ fn allowlist_account_works() {
...
+            EvmAccountAllowlist::<Test>::get(bab_id, EvmAddress::default()).unwrap(),
             MintStatus::Available(1)
         );
 
         assert_noop!(
             MantaSBTPallet::allowlist_evm_account(
                 MockOrigin::signed(ALICE),
-                EvmAddressType::Bab(EvmAddress::default())
+                bab_id,
+                EvmAddress::default(),
```
Comment (ghzlatarev) [inline pallets/manta-sbt/src/tests.rs:473]: is this default only for tests or is it actually needed in production ?
If it's not needed we should make this for tests only.

File: pallets/manta-sbt/src/lib.rs [hunk 3]
```diff
@@ -239,23 +279,57 @@ pub mod pallet {
...
+        MintId,
+        RegisteredMint<Moment<T>, T::RegistryBound>,
+        OptionQuery,
+    >;
+
     /// SBT Metadata maps `StandardAsset` to the correstonding SBT metadata
     ///
     /// Metadata is raw bytes that correspond to an image
+    ///
+    /// Deprecated delete after migration
```
Comment (ghzlatarev) [inline pallets/manta-sbt/src/lib.rs:332]: issue for this.

File: pallets/manta-sbt/src/lib.rs [hunk 4]
```diff
@@ -497,37 +573,128 @@ pub mod pallet {
...
 
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T>
+    where
+        T::AccountId: From<AccountId> + Into<AccountId>,
+    {
+        /// Migrates sbt pallet to new storage scheme
+        ///
+        /// remove once migration is complete
+        fn on_idle(_n: T::BlockNumber, remaining_weight: Weight) -> Weight {
```
Comment (ghzlatarev) [inline pallets/manta-sbt/src/lib.rs:656]: when is the on-idle actually executed.
after the extrinsics phase, if there's more weight available ?

**Review by ghzlatarev — COMMENTED** [review]
... didn't click submit on Saturday, so these were all pending.

File: pallets/manta-sbt/src/lib.rs (hunk 4 above)
Comment (ferrell-code) [reply pallets/manta-sbt/src/lib.rs:656]: IIRC, After the extrinsics and it is the last hook executed other than the `on_finalize` hook

File: pallets/manta-sbt/src/tests.rs (hunk 2 above)
Comment (ferrell-code) [reply pallets/manta-sbt/src/tests.rs:473]: I used default in tests where the value of `EvmAddress` was irrelevant.

I don't think there is a way to feature gate the Default trait as it originates from `H160` type. Shouldn't use default in prod

File: pallets/manta-sbt/src/lib.rs (hunk 3 above)
Comment (ferrell-code) [reply pallets/manta-sbt/src/lib.rs:332]: #1090

File: runtime/calamari/src/migrations/mod.rs (hunk 1 above)
Comment (ferrell-code) [reply runtime/calamari/src/migrations/mod.rs:18]: I think that is a good solution, could do in #1090

File: pallets/manta-sbt/src/lib.rs (hunk 4 above)
Comment (ferrell-code) [reply pallets/manta-sbt/src/lib.rs:656]: https://github.com/paritytech/substrate/blob/e699560d2037d155e6054c05b6334a6e9fa64926/frame/support/src/traits/hooks.rs#L77

File: pallets/manta-sbt/src/tests.rs (hunk 2 above)
Comment (ghzlatarev) [reply pallets/manta-sbt/src/tests.rs:473]: Oh it's not our own impl, ok then,

**Review by ghzlatarev — APPROVED** [review]

**Review by zqhxuyuan — APPROVED** [review]

**Pull request merged**